File: ffi/ts/src/rangeProof.ts
import {
  addRangeProofToVec,
  addToAmountRecoveryReqVec,
  addToUint64Vec,
  buildRangeProof,
  createAmountRecoveryReqVec,
  castToRangeProof,
  createRangeProofVec,
  createUint64Vec,
  deserializeRangeProof,
  freeAmountRecoveryReqVec,
  freeAmountsRetVal,
  freeObj,
  freeRangeProofVec,
  freeUint64Vec,
  genAmountRecoveryReq,
  getAmountRecoveryResultAmount,
  getAmountRecoveryResultIsSucc,
  getAmountRecoveryResultMsg,
  getAmountRecoveryResultSize,
  getRangeProof_A,
  getRangeProof_alpha_hat,
  getRangeProof_A_wip,
  getRangeProof_B,
  getRangeProof_delta_prime,
  getRangeProof_r_prime,
  getRangeProof_s_prime,
  getRangeProof_tau_x,
  recoverAmount,
  serializeRangeProof,
  verifyRangeProofs,
} from './blsct'

import { AmountRecoveryReq } from './amountRecoveryReq'
import { AmountRecoveryRes } from './amountRecoveryRes'
import { ManagedObj } from './managedObj'
import { Point } from './point'
import { Scalar } from './scalar'
import { TokenId } from './tokenId'

/** Represents a (possibly aggregated) range proof for one or more confidential transaction amounts.
 *
 * Examples:
 * ```ts
 * const { RangeProof, AmountRecoveryReq, AmountRecoveryRes, Point, TokenId } = require('navio-blsct')
 * const nonce = Point.random()
 * const tokenId = TokenId.default()
 * const rp = RangeProof.generate([456], nonce, 'navio', tokenId)
 * RangeProof.verifyProofs([rp]) // true
 * const req = new AmountRecoveryReq(rp, nonce)
 * const res = RangeProof.recoverAmounts([req1])
 * res[0].isSucc // true
 * res[0].amount // 456
 * res[0].message // 'navio'
 * rp.get_A() // Point object representing A
 * rp.get_A_wip() // Point object representing A_wip
 * rp.get_B() // Point object representing B
 * rp.get_r_prime() // Scalar object representing r'
 * rp.get_s_prime() // Scalar object representing s'
 * rp.get_delta_prime() // Scalar object representing delta'
 * rp.get_alpha_hat() // Scalar object representing alpha_hat
 * rp.get_tau_x() // Scalar object representing tau_x
 * const ser = rp.serialize()
 * const deser = RangeProof.deserialize(ser)
 * ser === deser.serialize() // true
 * ```
 */
export class RangeProof extends ManagedObj {
  constructor(obj: any) {
    super(obj)
  }

  /** Generates a range proof for the given amounts, nonce, and message.
   * @param amounts - An array of amounts to be included in the range proof.
   * @param nonce - A nonce used to generate the range proof.
   * @param msg - A message associated with the range proof.
   * @param tokenId - An optional token ID. If not provided, a default token ID is used.
   * @returns A new `RangeProof` instance containing the generated range proof.
   */
  static generate(
    amounts: number[],
    nonce: Point,
    msg: string,
    tokenId?: TokenId,
  ): RangeProof {
    tokenId = tokenId ?? TokenId.default()

    const vec = createUint64Vec()
    for (const amount of amounts) {
      addToUint64Vec(vec, amount)
    }

    const rv = buildRangeProof(
      vec,
      nonce.value(),
      msg,
      tokenId.value(),
    )
    freeUint64Vec(vec)

    if (rv.result !== 0) {
      const msg = `Building range proof failed. Error code = ${rv.result}`
      freeObj(rv)
      throw new Error(msg)
    }
    const x = RangeProof.fromObjAndSize(
      rv.value,
      rv.value_size,
    )
    freeObj(rv)
    return x
  }

  /** Verifies a list of range proofs.
   * @param proofs - An array of `RangeProof` instances to be verified.
   * @returns `true` if all proofs are valid, `false` otherwise.
   */
  static verifyProofs(proofs: RangeProof[]): boolean {
    const vec = createRangeProofVec()
    for (const proof of proofs) {
      addRangeProofToVec(vec, proof.size(), proof.value())
    }
    
    const rv = verifyRangeProofs(vec)
    if (rv.result !== 0) {
      const msg = `Verifying range proofs failed. Error code = ${rv.result}`
      freeObj(rv)
      throw new Error(msg)
    }
    freeRangeProofVec(vec)
    return rv.value
  }

  /** Recovers amounts from a list of `AmountRecoveryReq` instances.
   * @param reqs - An array of `AmountRecoveryReq` instances containing range proofs and nonces.
   * @returns An array of `AmountRecoveryRes` instances containing the recovery results.
   */
  recoverAmounts(reqs: AmountRecoveryReq[]): AmountRecoveryRes[] {
    const reqVec = createAmountRecoveryReqVec()

    for (const req of reqs) {
      const blsctReq = genAmountRecoveryReq(
        req.rangeProof.value(),
        req.rangeProof.size(),
        req.nonce.value(),
      )
      addToAmountRecoveryReqVec(reqVec, blsctReq)
    }

    const rv = recoverAmount(reqVec)
    freeAmountRecoveryReqVec(reqVec)

    if (rv.result !== 0) {
      const msg = `Recovering amount failed. Error code = ${rv.result}`
      freeAmountsRetVal(rv)
      throw new Error(msg)
    } 

    let results: AmountRecoveryRes[] = []
    const size = getAmountRecoveryResultSize(rv.value)

    for (let i=0; i<size; ++i) {
      const isSucc = getAmountRecoveryResultIsSucc(rv.value, i)
      const amount = getAmountRecoveryResultAmount(rv.value, i)
      const msg = getAmountRecoveryResultMsg(rv.value, i)
      const x = new AmountRecoveryRes(
        isSucc, 
        amount,
        msg,
      )
      results.push(x)
    }
    
    freeAmountsRetVal(rv)
    return results
  }

  override value(): any {
    return castToRangeProof(this.obj)
  }

  override serialize(): string {
    return serializeRangeProof(this.value(), this.size())
  }

  /** Deserializes a hexadecimal string into a `RangeProof` instance.
   * @param hex - A hexadecimal string representing the serialized range proof.
   * @returns A new `RangeProof` instance containing the deserialized data.
   */
  static deserialize(
    this: new (obj: any) => RangeProof,
    hex: string
  ): RangeProof {
    return RangeProof._deserialize(hex, deserializeRangeProof)
  }

  /** Returns the A point of the range proof.
    * @returns A `Point` object representing the A point.
    */
  get_A(): Point {
    const obj = getRangeProof_A(this.value(), this.size())
    return Point.fromObj(obj)
  }

  /** Returns the A_wip point of the range proof.
   * @returns A `Point` object representing the A_wip point.
   */
  get_A_wip(): Point {
    const obj = getRangeProof_A_wip(this.value(), this.size())
    return Point.fromObj(obj)
  }

  /** Returns the B point of the range proof.
   * @returns A `Point` object representing the B point.
   */
  get_B(): Point {
    const obj = getRangeProof_B(this.value(), this.size())
    return Point.fromObj(obj)
  }

  /** Returns the r' scalar of the range proof.
   * @returns A `Scalar` object representing the r' scalar.
   */
  get_r_prime(): Scalar {
    const obj = getRangeProof_r_prime(this.value(), this.size())
    return Scalar.fromObj(obj)
  }

  /** Returns the s' scalar of the range proof.
   * @returns A `Scalar` object representing the s' scalar.
   */
  get_s_prime(): Scalar {
    const obj = getRangeProof_s_prime(this.value(), this.size())
    return Scalar.fromObj(obj)
  }

  /** Returns the tau_x scalar of the range proof.
   * @returns A `Scalar` object representing the tau_x scalar.
   */
  get_delta_prime(): Scalar {
    const obj = getRangeProof_delta_prime(this.value(), this.size())
    return Scalar.fromObj(obj)
  }

  /** Returns the alpha_hat scalar of the range proof.
   * @returns A `Scalar` object representing the alpha_hat scalar.
   */
  get_alpha_hat(): Scalar {
    const obj = getRangeProof_alpha_hat(this.value(), this.size())
    return Scalar.fromObj(obj)
  }

  /** Returns the tau_x scalar of the range proof.
   * @returns A `Scalar` object representing the t_aux scalar.
   */
  get_tau_x(): Scalar {
    const obj = getRangeProof_tau_x(this.value(), this.size())
    return Scalar.fromObj(obj)
  }
}

File: ffi/rust/src/range_proof.rs
use crate::{
  amount_recovery_req::AmountRecoveryReq,
  amount_recovery_res::AmountRecoveryRes,
  blsct_obj::{BlsctObj, self},
  blsct_serde::BlsctSerde, 
  ffi::{
    add_to_amount_recovery_req_vec,
    add_to_uint64_vec,
    add_to_range_proof_vec,
    BlsctPoint,
    BlsctRangeProof,
    BlsctRetVal,
    BlsctScalar,
    build_range_proof,
    create_amount_recovery_req_vec,
    create_range_proof_vec,
    create_uint64_vec,
    delete_amount_recovery_req_vec,
    delete_range_proof_vec,
    delete_uint64_vec,
    deserialize_range_proof,
    free_obj,
    free_amounts_ret_val,
    gen_amount_recovery_req,
    get_amount_recovery_result_amount,
    get_amount_recovery_result_is_succ,
    get_amount_recovery_result_msg,
    get_amount_recovery_result_size,
    get_range_proof_A,
    get_range_proof_A_wip,
    get_range_proof_B,
    get_range_proof_r_prime,
    get_range_proof_s_prime,
    get_range_proof_delta_prime,
    get_range_proof_alpha_hat,
    get_range_proof_tau_x,
    recover_amount,
    serialize_range_proof,
    verify_range_proofs,
  },
  macros::{
    impl_clone,
    impl_display,
    impl_from_retval,
    impl_size,
    impl_value,
  },
  point::Point,
  scalar::Scalar,
  token_id::TokenId,
};
use serde::{Deserialize, Serialize};
use std::{
  ffi::{
    CStr,
    CString,
    c_char,
    c_void,
    NulError,
  },
  fmt,
};

#[derive(Debug, PartialEq, Eq)]
pub enum Error<'a> {
  BlsctObjError(blsct_obj::Error<'a>),
  FailedToCreateUint64Vec,
  FailedToCreateCString(NulError),
  FailedToCreateRangeProofVector,
  FailedToVerifyRangeProofs(u8),
  FailedToCreateAmountRecoveryRequestVector,
  FailedToRecoverAmount(u8),
}

impl<'a> std::error::Error for Error<'a> {}

impl<'a> fmt::Display for Error<'a> {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      Error::BlsctObjError(e) =>
        write!(f, "BlsctObjError: {e:?}"),
      Error::FailedToCreateUint64Vec =>
        write!(f, "Failed to create uint64_t vector"),
      Error::FailedToCreateCString(e) =>
        write!(f, "Failed to create CString: {e:?}"),
      Error::FailedToCreateRangeProofVector =>
        write!(f, "Failed to create range proof vector"),
      Error::FailedToVerifyRangeProofs(e) =>
        write!(f, "Failed to verify range proofs: {e}"),
      Error::FailedToCreateAmountRecoveryRequestVector =>
        write!(f, "Failed to create amount recovery request vector"),
      Error::FailedToRecoverAmount(e) =>
        write!(f, "Failed to recover amount: {e}"),
    }
  }
}

#[derive(Debug, Deserialize, Serialize, Eq)]
pub struct RangeProof {
  obj: BlsctObj<RangeProof, BlsctRangeProof>,
}

impl_from_retval!(RangeProof);
impl_display!(RangeProof);
impl_clone!(RangeProof);

impl RangeProof {
  pub fn new<'a>(
    amounts: &Vec<u64>,
    nonce: &Point,
    msg: &str,
    token_id: &TokenId
  ) -> Result<Self, Error<'a>> {
    let vp_u64_vec = {
      let vec = unsafe { create_uint64_vec() };
      if vec.is_null() {
        return Err(Error::FailedToCreateUint64Vec);
      }
      for amount in amounts {
        unsafe { add_to_uint64_vec(vec, *amount) };
      }
      vec
    };
    let c_msg = CString::new(msg)
      .map_err(|e| Error::FailedToCreateCString(e))?;

    let rv = unsafe { build_range_proof(
      vp_u64_vec,
      nonce.value(),
      c_msg.as_ptr(),
      token_id.value(),
    )};
    unsafe { delete_uint64_vec(vp_u64_vec) };

    let obj = BlsctObj::from_retval(rv)
      .map_err(|e| Error::BlsctObjError(e))?;
    Ok(obj.into())
  }

  pub fn verify_proofs<'a>(proofs: &Vec<RangeProof>) -> Result<bool, Error<'a>> {
    let range_proofs = unsafe { create_range_proof_vec() };
    if range_proofs.is_null() {
      return Err(Error::FailedToCreateRangeProofVector);
    }

    for proof in proofs {
      let proof_size = proof.obj.size();
      unsafe { add_to_range_proof_vec(
        range_proofs,
        proof.value(),
        proof_size,
      )};
    }

    let rv = unsafe { verify_range_proofs(range_proofs) };
    let (result, value) = unsafe { ((*rv).result, (*rv).value) };
    unsafe { 
      delete_range_proof_vec(range_proofs);
      free_obj(rv as *mut c_void);
    };

    if result == 0 { 
      Ok(value)
    } else {
      Err(Error::FailedToVerifyRangeProofs(result))
    }
  }

  pub fn recover_amounts<'a>(
    reqs: Vec<AmountRecoveryReq>,
  ) -> Result<Vec<AmountRecoveryRes>, Error<'a>> {

    let req_vec = unsafe { create_amount_recovery_req_vec() };
    if req_vec.is_null() {
      return Err(Error::FailedToCreateAmountRecoveryRequestVector);
    }
    for req in reqs {
      let blsct_req = unsafe { gen_amount_recovery_req(
        req.range_proof.value() as *mut c_void,
        req.range_proof.size(),
        req.nonce.value() as *mut c_void,
      )};

      unsafe { add_to_amount_recovery_req_vec(
        req_vec,
        blsct_req as *mut c_void,
      )};
    }

    let rv = unsafe { recover_amount(req_vec) };
    unsafe { delete_amount_recovery_req_vec(req_vec) };

    let (result, value) = unsafe { ((*rv).result, (*rv).value) };
    if result != 0 {
      unsafe { free_amounts_ret_val(rv) };
      return Err(Error::FailedToRecoverAmount(result));
    }

    let mut results: Vec<AmountRecoveryRes> = vec![];
    let result_size = unsafe { get_amount_recovery_result_size(value) } as usize;

    for i in 0..result_size {
      let is_succ = unsafe { get_amount_recovery_result_is_succ(value, i) };
      let amount = unsafe { get_amount_recovery_result_amount(value, i) };
      let msg_c_str = unsafe { get_amount_recovery_result_msg(value, i) };
      let msg = if msg_c_str.is_null() { "" } else {
        unsafe { CStr::from_ptr(msg_c_str) }.to_str().unwrap()
      };
      let result = AmountRecoveryRes::new(
        is_succ,
        amount,
        &msg,
      );
      results.push(result);
    }

    Ok(results)
  }

  #[allow(non_snake_case)]
  pub fn get_A(&self) -> Point {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_A(self.value(), size) };
    BlsctObj::<Point, BlsctPoint>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_A_wip(&self) -> Point {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_A_wip(self.value(), size) };
    BlsctObj::<Point, BlsctPoint>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_B(&self) -> Point {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_B(self.value(), size) };
    BlsctObj::<Point, BlsctPoint>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_r_prime(&self) -> Scalar {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_r_prime(self.value(), size) };
    BlsctObj::<Scalar, BlsctScalar>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_s_prime(&self) -> Scalar {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_s_prime(self.value(), size) };
    BlsctObj::<Scalar, BlsctScalar>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_delta_prime(&self) -> Scalar {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_delta_prime(self.value(), size) };
    BlsctObj::<Scalar, BlsctScalar>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_alpha_hat(&self) -> Scalar {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_alpha_hat(self.value(), size) };
    BlsctObj::<Scalar, BlsctScalar>::from_c_obj(obj).into()
  }

  #[allow(non_snake_case)]
  pub fn get_tau_x(&self) -> Scalar {
    let size = self.obj.size();
    let obj = unsafe { get_range_proof_tau_x(self.value(), size) };
    BlsctObj::<Scalar, BlsctScalar>::from_c_obj(obj).into()
  }

  impl_size!();
  impl_value!(BlsctRangeProof);
}

impl BlsctSerde for RangeProof {
  unsafe fn serialize(ptr: *const u8, size: usize) -> *const i8 {
    serialize_range_proof(ptr as *const BlsctRangeProof, size)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    let c_str = unsafe { CStr::from_ptr(hex) };
    let obj_size = c_str.to_bytes().len() / 2;
    deserialize_range_proof(hex, obj_size)
  }
}

impl From<BlsctObj<RangeProof, BlsctRangeProof>> for RangeProof {
  fn from(obj: BlsctObj<RangeProof, BlsctRangeProof>) -> RangeProof {
    RangeProof { obj }
  }
}

impl PartialEq for RangeProof {
  fn eq(&self, other: &Self) -> bool {
    self.get_A() == other.get_A() &&
    self.get_B() == other.get_B() &&
    self.get_r_prime() == other.get_r_prime() &&
    self.get_s_prime() == other.get_s_prime() &&
    self.get_delta_prime() == other.get_delta_prime() &&
    self.get_alpha_hat() == other.get_alpha_hat() &&
    self.get_tau_x() == other.get_tau_x()
  }
}


#[cfg(test)]
mod tests {
  use super::*;
  use crate::initializer::init;

  fn gen_range_proof() -> RangeProof {
    let values = vec![123u64];
    let nonce = Point::random().unwrap();
    let token_id = TokenId::default().unwrap();
    RangeProof::new(&values, &nonce, "navio", &token_id).unwrap()
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_A() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_A();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_A_wip() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_A_wip();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_B() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_B();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_r_prime() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_r_prime();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_s_prime() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_s_prime();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_delta_prime() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_delta_prime();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_alpha_hat() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_alpha_hat();
  }

  #[test]
  #[allow(non_snake_case)]
  fn test_get_tau_x() {
    init();
    let rp = gen_range_proof();
    let _ = rp.get_tau_x();
  }

  #[test]
  fn test_prove_and_verify() {
    init();
    let rp = gen_range_proof();
    let result = RangeProof::verify_proofs(&vec![rp]).unwrap();
    assert!(result);
  }

  #[test]
  fn test_recover_amounts() {
    init();

    let msg = "navio";
    let amount = 123u64;

    let values = vec![amount];
    let nonce = Point::random().unwrap();
    let token_id = TokenId::default().unwrap();

    let rp = RangeProof::new(&values, &nonce, msg, &token_id).unwrap();

    let req = AmountRecoveryReq::new(&rp, &nonce);
    let res = RangeProof::recover_amounts(vec![req]).unwrap();

    assert_eq!(res.len(), 1);
    assert_eq!(res[0].is_succ, true);
    assert_eq!(res[0].amount, amount);
    assert_eq!(res[0].msg, msg);
  }

  #[test]
  fn test_deser() {
    init();
    let a = gen_range_proof();
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<RangeProof>(&hex).unwrap();
    assert_eq!(a, b);
  }
}

